system/jlib/jutil.cpp

```diff
@@ -3030,19 +3051,117 @@ int getEnum(const char *v, const EnumMapping *map, int defval)
 
 //---------------------------------------------------------------------------------------------------------------------
 
+static void splitUrlAddress(const char *address, size_t len, StringBuffer &host, StringBuffer *port)
+{
+    if (isEmptyString(address)||len==0)
+        return;
+    const char *sep = (const char *)memchr(address, ':', len);
+    if (!sep)
+        host.append(len, address);
+    else
+    {
+        host.append(sep - address, address);
+        len = len - (sep - address) - 1;
+        if (port)
+            port->append(len, sep+1);
+        else
+            host.append(':').append(len, sep+1);
+    }
+}
+
+static void splitUrlAuthority(const char *authority, size_t authorityLen, StringBuffer &user, StringBuffer &password, StringBuffer &host, StringBuffer *port)
+{
+    if (isEmptyString(authority)||authorityLen==0)
+        return;
+    const char *at = (const char *) memchr(authority, '@', authorityLen);
+    if (!at)
+        splitUrlAddress(authority, authorityLen, host, port);
+    else
+    {
+        size_t userinfoLen = (at - authority);
+        splitUrlAddress(at+1, authorityLen - userinfoLen - 1, host, port);
+        const char *sep = (const char *) memchr(authority, ':', at - authority);
+        if (!sep)
+            user.append(at-authority, authority);
+        else
+        {
+            user.append(sep-authority, authority);
+            size_t passwordLen = (at - sep - 1);
+            password.append(passwordLen, sep+1);
+        }
+    }
+}
+
+static inline void extractUrlProtocol(const char *&url, bool &https, StringBuffer *scheme)
+{
+    if(!url || strlen(url) <= 7)
+        throw MakeStringException(-1, "Invalid URL %s", url);
+    if (0 == strncasecmp(url, "HTTPS://", 8))
+        https = true;
+    else if (0 == strncasecmp(url, "HTTP://", 7))
```

Comment:
question: are there likely to be other protocols?

Reply:
In terms of vaults I would think most are HTTPS/HTTP accessible.  In terms of general future thinking I could see us supporting protocol buffers / gRPC in the future.  I could remove the https flag for now, since I don't use it.  Or change is to an ssl flag.  I was trying to make functions that might be useful elsewhere.